cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh

```diff
@@ -0,0 +1,862 @@
+    k_th_ = shfl(val_arr_[max_arr_len_ - 1], k_th_lane_);
+  }
+
+  __device__ void merge_buf_()
+  {
+    bitonic_sort<max_buf_len_ * WarpSize, greater>::run(val_buf_, idx_buf_);
+
+    if (max_arr_len_ > max_buf_len_) {
+      for (int i = 0; i < max_buf_len_; ++i) {
+        T& val = val_arr_[max_arr_len_ - max_buf_len_ + i];
+        T& buf = val_buf_[i];
+        if (is_greater_than<greater>(buf, val)) {
+          val                                       = buf;
+          idx_arr_[max_arr_len_ - max_buf_len_ + i] = idx_buf_[i];
+        }
+      }
+    } else if (max_arr_len_ < max_buf_len_) {
+      for (int i = 0; i < max_arr_len_; ++i) {
+        T& val = val_arr_[i];
+        T& buf = val_buf_[max_buf_len_ - max_arr_len_ + i];
+        if (is_greater_than<greater>(buf, val)) {
+          val         = buf;
+          idx_arr_[i] = idx_buf_[max_buf_len_ - max_arr_len_ + i];
+        }
+      }
+    } else {
+      for (int i = 0; i < max_arr_len_; ++i) {
+        if (is_greater_than<greater>(val_buf_[i], val_arr_[i])) {
+          val_arr_[i] = val_buf_[i];
+          idx_arr_[i] = idx_buf_[i];
+        }
+      }
+    }
+
+    bitonic_merge<capacity, !greater>::run(val_arr_, idx_arr_);
+
+    buf_len_ = 0;
+    set_k_th_();  // contains sync
+    for (int i = 0; i < max_buf_len_; ++i) {
+      val_buf_[i] = dummy_;
+    }
+  }
+
+  using WarpSort<capacity, greater, T, IdxT>::max_arr_len_;
+  using WarpSort<capacity, greater, T, IdxT>::val_arr_;
+  using WarpSort<capacity, greater, T, IdxT>::idx_arr_;
+  using WarpSort<capacity, greater, T, IdxT>::lane_;
+  using WarpSort<capacity, greater, T, IdxT>::k_;
+  using WarpSort<capacity, greater, T, IdxT>::dummy_;
+
+  static constexpr int max_buf_len_ = (capacity <= 64) ? 2 : 4;
+
+  T val_buf_[max_buf_len_];
+  IdxT idx_buf_[max_buf_len_];
+  int buf_len_;
+
+  T k_th_;
+  const int k_th_lane_;
+};
+
+template <int capacity, bool greater, typename T, typename IdxT>
+class WarpBitonic : public WarpSort<capacity, greater, T, IdxT> {
+ public:
+  __device__ WarpBitonic(int k, T dummy)
+    : WarpSort<capacity, greater, T, IdxT>(k, dummy), buf_len_(0)
+  {
+    for (int i = 0; i < max_arr_len_; ++i) {
+      val_buf_[i] = dummy_;
+    }
+  }
+
+  __device__ void add(const T* in, IdxT start, IdxT end)
+  {
+    add_first_(in, start, end);
+    start += capacity;
+    while (start < end) {
+      add_extra_(in, start, end);
+      merge_();
+      start += capacity;
+    }
+  }
+
+  __device__ void add(const T* in, const IdxT* in_idx, IdxT start, IdxT end)
+  {
+    add_first_(in, in_idx, start, end);
+    start += capacity;
+    while (start < end) {
+      add_extra_(in, in_idx, start, end);
+      merge_();
+      start += capacity;
+    }
+  }
+
+  __device__ void add(T val, IdxT idx)
+  {
+    for (int i = 0; i < max_arr_len_; ++i) {
+      if (i == buf_len_) {
+        val_buf_[i] = val;
+        idx_buf_[i] = idx;
+      }
+    }
```

Comment:
Apparently, the loop is unrolled, which ensures the constant indexing of the arrays. When directly indexing using `buf_len_`, the array is placed in local memory instead of the registers, which makes the whole algorithm slower (~2x slower on my machine).

Reply:
Ok, then let us keep the loop how it is.